apis/placement/v1alpha1/stagedupdate_types.go

```diff
@@ -0,0 +1,481 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"go.goms.io/fleet/apis/placement/v1beta1"
+)
+
+// +genclient
+// +genclient:namespaced
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope="Namespaced",categories={fleet,fleet-placement}
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// StagedUpdateRun defines a stage by stage update run that applies the selected resources by the
+// corresponding ClusterResourcePlacement to its selected clusters. We remove the resources from the clusters that are
+// unselected after all the stages explicitly defined in the updateStrategy complete.
+// Each StagedUpdateRun object corresponds to a single "release" of a certain version of the resources.
+// The release is abandoned if the StagedUpdateRun object is deleted or the scheduling decision (i.e., the selected clusters) changes.
+// The name of the StagedUpdateRun needs to conform to [RFC 1123](https://tools.ietf.org/html/rfc1123).
```

Comment:
Can we make sure the names are shorter so that in the future if we need to have an approval name like <runname>-<stagename>-<group-name>-<cluster-name>, it is shorter than 253 characters. We can keep each name to be shorter than 32 (we can relax the limit in the future if needed).

Reply:
rfc 1123 limit is 63 so one can concatenate 4 parts into the approval name.